From 76369d8c7e2ed1f475ffe10771d783c8d6ea3412 Mon Sep 17 00:00:00 2001
From: Stefan Haller <stefan@haller-berlin.de>
Date: Wed, 18 Apr 2012 14:33:35 +0200
Subject: [PATCH 07/20] gitk: Make svn revision numbers clickable links

---
 gitk-git/gitk | 35 ++++++++++++++++++++++++++++++++---
 1 file changed, 32 insertions(+), 3 deletions(-)

diff --git a/gitk-git/gitk b/gitk-git/gitk
index f1c28a79ed..10e80139d3 100755
--- a/gitk-git/gitk
+++ b/gitk-git/gitk
@@ -7081,8 +7081,26 @@ proc appendwithlinks {text tags} {
 
     set start [$ctext index "end - 1c"]
     $ctext insert end $text $tags
-    set links [regexp -indices -all -inline {(?:\m|-g)[0-9a-f]{6,40}\M} $text]
-    foreach l $links {
+    set links [regexp -expanded -indices -all -inline {
+        # subversion revision footer line
+          svn:\s[\w.-]{3,40}@[1-9][0-9]{0,4}\M
+        # 5 digit number prepended with 'r' or '@' (g1)
+        | (?:[r@])([1-9][0-9]{0,4})\M
+        # exactly 5 digit number (g2)
+        | ([1-9][0-9]{4})\M
+        # 6 to 40 digit sha 1 (g3)
+        | (?:\m|-g)([0-9a-f]{6,40})\M
+    } $text]
+    foreach {all g1 g2 g3} $links {
+        if {[lindex $g1 0] != -1} {
+            set l $g1
+        } elseif {[lindex $g2 0] != -1} {
+            set l $g2
+        } elseif {[lindex $g3 0] != -1} {
+            set l $g3
+        } else {
+            continue
+        }
         set s [lindex $l 0]
         set e [lindex $l 1]
         set linkid [string range $text $s $e]
@@ -7115,7 +7133,11 @@ proc setlink {id lk} {
     }
 
     set known 0
-    if {[string length $id] < 40} {
+    set len [string length $id]
+    if {$len < 6} {
+        # always assume subversion revisions to exist
+        set known 1
+    } elseif {$len < 40} {
         set matches [longid $id]
         if {[llength $matches] > 0} {
             if {[llength $matches] > 1} return
@@ -8987,6 +9009,13 @@ proc normalline {} {
 }
 
 proc selbyid {id {isnew 1}} {
+    if {[string length $id] < 6} {
+        global findstring gdttype
+        set gdttype "containing:"
+        set findstring "@$id"
+        dofind
+        return
+    }
     global curview
     if {[commitinview $id $curview]} {
         selectline [rowofcommit $id] $isnew
-- 
2.40.1

